jax/experimental/fused_attention_stableHLO_test.py

```diff
@@ -0,0 +1,124 @@
+    bias = jax.random.uniform(key2, (4, args.batch_size, args.num_attn_heads, args.q_seq_len, args.kv_seq_len), dtype=dtype) if args.bias else None
+    mask = jax.random.uniform(key2, (4, args.batch_size, args.num_attn_heads, args.q_seq_len, args.kv_seq_len), dtype=dtype) if args.mask else None
+    grad = jax.random.uniform(key2, (4, args.batch_size, args.q_seq_len, args.num_attn_heads, args.head_dim), dtype=dtype)
+
+    if args.fwd_only:
+        jitted_fwd_step = jax.jit(partial(f, mask_type=args.mask_type, scale=args.scale))
+        out = jitted_fwd_step(input_q, input_k, input_v, bias, mask)
+        return out
+    else:
+        jitted_train_step = jax.jit(train_step)
+        # out, input_grad = jitted_train_step(args.mask_type, args.scale, input_q, input_k, input_v, bias, mask, grad)
+        # out, input_grad = jitted_train_step(input_q, input_k, input_v, bias, mask, grad)
+        # print(input_grad[0][0,0,0,:20])
+        # return out, input_grad
+        import numpy as np
+        from jax.sharding import Mesh
+        from jax.sharding import PartitionSpec, NamedSharding
+        from jax.experimental.pjit import pjit
+        devices = np.array(jax.local_devices())
+        devices = devices.reshape((2, 4, 1, 1, 1))
+        with Mesh(devices, ('p', 'b', 's', 'n', 'd')) as mesh:
+            input_q = jax.device_put(input_q, NamedSharding(mesh, PartitionSpec('p', 'b', None, None, None)))
+            input_k = jax.device_put(input_k, NamedSharding(mesh, PartitionSpec('p', 'b', None, None, None)))
+            input_v = jax.device_put(input_v, NamedSharding(mesh, PartitionSpec('p', 'b', None, None, None)))
+            bias = jax.device_put(bias, NamedSharding(mesh, PartitionSpec('p', 'b', None, None, None)))
+            mask = jax.device_put(mask, NamedSharding(mesh, PartitionSpec('p', 'b', None, None, None)))
+            grad = jax.device_put(grad, NamedSharding(mesh, PartitionSpec('p', 'b', None, None, None)))
+
+            pjitter = pjit(jitted_train_step,
+                in_shardings=(
+                            PartitionSpec('p', 'b', None, None, None),
+                            PartitionSpec('p', 'b', None, None, None),
+                            PartitionSpec('p', 'b', None, None, None),
+                            PartitionSpec('p', 'b', None, None, None),
+                            PartitionSpec('p', 'b', None, None, None),
+                            PartitionSpec('p', 'b', None, None, None)),
+                out_shardings=(None, 
+                            (PartitionSpec('p', 'b', None, None, None),
+                            PartitionSpec('p', 'b', None, None, None),
+                            PartitionSpec('p', 'b', None, None, None),
+                            None, 
+                            None))
+            )
+
+            out, grads = pjitter(input_q, input_k, input_v, bias, mask, grad)
+            print(grads[0][0,0,0,0,:20])
+
+if __name__ == "__main__":
+    main()
```

Comment:
Should we call it with/wo bias/mask/dropout and forward only? Just to test all the combination here.
This will also need sharding tests. You can look at https://github.com/NVIDIA/TransformerEngine/blob/main/tests/jax/test_distributed_fused_attn.py for an example.

Reply:
Are you use this file end up tested if we do:
`pytest jax/experimental`? I don't think so, so it won't be executed by the CI.

Reply:
This is just for an example, I will add more tests in the future.